core/plannernotes.c

```diff
@@ -30,16 +30,65 @@ int diveplan_duration(struct diveplan *diveplan)
 	return (duration + 30) / 60;
 }
 
+struct icd_data { // This structure provides communication between function isobaric_counterdiffusion() and the calling software.
+	int time; 			// For a trimix gas change, the time (in seconds) of the change
+	int ambientpressure_mbar;	// The ambient pressure (mbar) when the gas change took place
+	int dN2;			// The change in fraction (permille) of nitrogen during the change
+	int dHe;			// The change in fraction (permille) of helium during the change
+};
+
+/* Perform isobaric counterdiffusion calculations for gas changes in trimix dives.
+ * Here we use the rule-of-fifths where, during a change involving trimix gas, the increase in nitrogen
+ * should not exceed one fifth of the decrease in helium.
+ * Parameters:	1) Pointer to the dive structure.
+ *				2) Pointer to the dive planner point where the gas change is taking place.
+ *				3) pointers to two gas mixes, the gas being switched from and the gas being switched to.
+ * 				4) a pointer to an icd_data structure.
+ * Output:		 i) The icd_data stucture is filled with the time, pressure, delta_N2 and delta_He numbers (as permille).
+ * 				ii) Function returns a boolean indicating an exceeding of the rule-of-fifths. False = no problem.
+ */
+bool isobaric_counterdiffusion(struct dive *dive, struct divedatapoint *dp, struct gasmix *oldgasmix, struct gasmix *newgasmix, struct icd_data *results)
+{
+	results->time = dp->time;
+	results->ambientpressure_mbar = depth_to_mbar(dp->depth.mm, dive);
+	results->dN2 = get_he(oldgasmix) + get_o2(oldgasmix) - get_he(newgasmix) - get_o2(newgasmix);
+	results->dHe = get_he(newgasmix) - get_he(oldgasmix);
+	return (get_he(oldgasmix) && (results->dN2 > 0) && ((5 * results->dN2) > -results->dHe));
+}
+
+/* Add the icd results of one trimix gas change to the dive plan html buffer. Two rows are added to the table, one
+ * indicating fractions of gas, the other indication partial pressures of gas. This function makes use of the
+ * icd_data structure that was filled with information by the function isobaric_counterdiffusion().
+ * Parameters:	1) Information about the buffer to output text to: *icdbuffer, bufsize, icdlen.
+ *				2) The data structure containing icd calculation results: icdvalues.
+ *				3) The names of the gasses in the gas change: gas-from and gas-to.
+ * Returns:		The size of the output buffer that has been used after the new results have been added (icdlen).
+ */
+int add_icd_entry(char *icdbuffer, unsigned int bufsize, int icdlen, struct icd_data icdvalues, char *gasname_from, const char *gasname_to)
```

Comment:
Is there a reason you passion icd_data by value rather than by pointer?
You could omit passing in icdlen and rather pass in icdbuffer+icdlen and bufsize-icdlen and return the increase.

Reply:
done. icd_data is passed as pointer.
the first three arguments have now been educed to 2 arguments as suggested.